### DISCUSSION TITLE
How to achieve soft keywords?

### DISCUSSION THREAD

**Discussion by ymc9** [eclipse-langium/langium#1012] [Q&A] [answered]
Hi,

My Langium schema contains keywords that are contextual. I.e., only has special meaning when used in an expression. For example: `X in Y` is an InExpression, but I don't want "in" to be an unconditional keyword, so I can use it in other contexts as a regular ID.

Is there a way to achieve this in Langium?

**Comment by spoenemann** [comment, marked as answer]
The usual solution is to use a "data type rule" for regular IDs:
```
RegularID returns string: ID | 'in';
```
That includes the listed keywords where a regular ID is possible. However, that's not considered in syntax highlighting, so `in` would still be highlighted as keyword regardless where it's used. To avoid that, you can turn off generating the highlighting file and manually remove the keyword.

**Reply by msujew** [reply]
Note that the `in` keyword can later still be highlighted as a keyword using the `SemanticTokenProvider` LSP service. Or in reverse, the `in` identifier can be highlighted as a variable, even though it is declared as a keyword in the syntactic token provider (textMate).

**Comment by ymc9** [comment]
Thank you! I found I've already done this in another place, just wasn't sure if it's the right solution 😄.

Yes, the default highlighting behavior is not ideal. I'll look into the solution.

🙏
